apollo-coroutines-support/src/main/kotlin/com/apollographql/apollo/coroutines/CoroutinesExtensions.kt

```diff
@@ -135,7 +142,17 @@ fun <T> ApolloSubscriptionCall<T>.toChannel(capacity: Int = Channel.UNLIMITED):
         }
 
         override fun onResponse(response: Response<T>) {
-            channel.offer(response)
+            try {
+                channel.offer(response)
+            } catch (ex: ClosedSendChannelException) {
+                // Swallow exception if the channel has already been closed without a cause.
+                // Due to concurrency, this can be called between channel.close and cancel().
+                // For the same reason, we cannot use channel.isClosedForSend.
+            } catch (ex: ApolloException) {
+                // Swallow exception if the channel has already been closed with an ApolloException.
+                // Due to concurrency, this can be called between channel.close and cancel().
+                // For the same reason, we cannot use channel.isClosedForSend.
```

Comment:
So the scenario would be?

* `onFailure(apolloSubscriptionTerminatedException)` is called, that closes the channel.
* `onResponse(response)` is called. We try to offer on a now closed channel.

That contradicts https://github.com/apollographql/apollo-android/blob/9a0e00364c3d52e4f0ae7fc5f17c274f11a32d4f/apollo-runtime/src/main/java/com/apollographql/apollo/ApolloSubscriptionCall.java#L67: onResponse should not be called after onFailure. Or is there another path leading here ?

Reply:
I didn't find how this is possible but we had that kind of crash:
```
com.apollographql.apollo.exception.ApolloNetworkException: 
  at com.apollographql.apollo.internal.RealApolloSubscriptionCall$SubscriptionManagerCallback.onNetworkError (RealApolloSubscriptionCall.java:150)
  at com.apollographql.apollo.internal.subscription.RealSubscriptionManager$SubscriptionRecord.notifyOnNetworkError (RealSubscriptionManager.java:409)
  at com.apollographql.apollo.internal.subscription.RealSubscriptionManager.onTransportFailure (RealSubscriptionManager.java:226)
  at com.apollographql.apollo.internal.subscription.RealSubscriptionManager$SubscriptionTransportCallback$2.run (RealSubscriptionManager.java:439)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at java.lang.Thread.run (Thread.java:764)
Caused by: java.io.EOFException: 
  at okio.RealBufferedSource.require (RealBufferedSource.java:61)
  at okio.RealBufferedSource.readByte (RealBufferedSource.java:74)
  at okhttp3.internal.ws.WebSocketReader.readHeader (WebSocketReader.java:117)
  at okhttp3.internal.ws.WebSocketReader.processNextFrame (WebSocketReader.java:101)
  at okhttp3.internal.ws.RealWebSocket.loopReader (RealWebSocket.java:274)
  at okhttp3.internal.ws.RealWebSocket$2.onResponse (RealWebSocket.java:214)
  at okhttp3.RealCall$AsyncCall.execute (RealCall.java:203)
  at okhttp3.internal.NamedRunnable.run (NamedRunnable.java:32)
```

Reply:
What about:

```kotlin
            try {
                channel.offer(response)
            } catch (ex: ClosedSendChannelException) {
                // Swallow exception if the channel has already been closed.
            } catch (ex: CancellationException) {
                // Swallow exception if the channel has already been cancelled.
            } catch (ex: ApolloException) {
                throw Exception("onResponse called after onFailure", ex)
           }
```

This will most likely not fix your issue in the last case but it will give more information where the problem might come from ?